[receiver/filelog] Add documentation about contents of the offset tracking file (#31886)
open-telemetry/opentelemetry-collector-contrib: receiver/filelogreceiver/README.md

## Offset tracking

The `storage` setting allows you to define the proper storage extension for storing file offsets.
While the storage parameter can ensure that log files are consumed accurately, it is possible that
logs are dropped while moving downstream through other components in the collector.
For additional resiliency, see [Fault tolerant log collection example](../../examples/fault-tolerant-logs-collection/README.md)

Here is some of the information the `filelog` receiver stores:
- The number of files it is currently tracking (`knownFiles`).
- For each file being tracked:

>   - The fingerprint of the file (`Fingerprint.first_bytes`).

Reviewer: Not sure if we have anywhere else documented the fingerprint logic but maybe you can point to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/fileconsumer/design.md#fingerprints for those that are interested to learn more about how the fingerprint is calculated and handled.
Otherwise we can move the content here 🤔 ?

Author: Yep, good idea, will do.

Author: Done in 08405ade31.

  - The byte offset from the start of the file, indicating the position in the file from where the

>     `filelog` receiver will continue reading the file (`Offset`).

Reviewer 2: ```suggestion
    `filelog` receiver continues reading the file (`Offset`).
```

>   - An arbitrary set of file attributes, e.g. the name of the file (`FileAttributes`).

Reviewer 2: ```suggestion
  - An arbitrary set of file attributes, such as the name of the file (`FileAttributes`).
```

## Offset tracking

The `storage` setting allows you to define the proper storage extension for storing file offsets.
While the storage parameter can ensure that log files are consumed accurately, it is possible that
logs are dropped while moving downstream through other components in the collector.
For additional resiliency, see [Fault tolerant log collection example](../../examples/fault-tolerant-logs-collection/README.md)

> ### File storage

Reviewer: Since this addition essentially documents the `file_storage` extension, wouldn't fit better at https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage/filestorage?
Then we could link from here the respective part with something like "from more info about the `file_storage` interanals check <link>".

Also, that would allow us to link from [`Fault tolerant log collection example`](../../examples/fault-tolerant-logs-collection/README.md) as well.

@ycombinator @djaglowski wdyt?

Reviewer 3: I agree. I'd rather not have this receiver get into documentation of other components.

Author: This addition documents how `filelog` receiver offsets are stored using the `file_storage` extension.  I think the useful parts of this addition are the details pertaining to the data persisted by the `filelog` receiver (signature, offset, etc.).  So I think this documentation belongs with the `filelog` receiver itself, not the `file_storage` extension.

Author: Yeah, I like that separation, @djaglowski.  Thanks for suggesting it. Will do!

Author: Documented the information persisted by the `filelog` receiver in 8211b78ea2370622053afb1960333a3bbe78459d.  I'll put up a separate PR for the `file_storage ` extension to describe how to inspect the contents of the bbolt data store.

Author: Documented the information persisted by the `filelog` receiver in 8211b78ea2370622053afb1960333a3bbe78459d.  I'll put up a separate PR for the `file_storage ` extension to describe how to inspect the contents of the bbolt data store.

Author: > I'll put up a separate PR for the `file_storage ` extension to describe how to inspect the contents of the bbolt data store.

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32180

> The [`filestorage` extension](../../extension/storage/filestorage) is a common storage extension that's used for tracking log file offsets. Sometimes, typically for debugging reasons, it's useful to view the file in which offsets are stored. The simplest way to do this is by printing out the contents of this file using the `strings` utility.

Reviewer: nit: isn't the name that we use in the pipelines `file_storage`? Using `filestorage` here might misslead to wrong configurations :)

> The following configuration shows a collector pipeline that's using the `filelog` receiver with the `storage` extension. Note that [compaction](../../extension/storage/filestorage/README.md#compaction) is not being used.

Reviewer: ```suggestion
The following configuration shows a collector pipeline that's using the `filelog` receiver with the `file_storage` extension. Note that [compaction](../../extension/storage/filestorage/README.md#compaction) is not being used.
```
`storage` is technically a setting (not an extension) that links to the storage extension, right?

the contents of this file using the `strings` utility.

Consider a collector pipeline that's using the `filelog` receiver with the `storage` extension as shown
below. Note that [compaction](../../extension/storage/filestorage/README.md#compaction) is not being used.

```yaml
receivers:
  filelog:
    include: /tmp/*.log
    storage: file_storage/filelogreceiver

exporters:
  ...

extensions:
  file_storage/filelogreceiver:
    directory: /tmp/otelcol/file_storage/filelogreceiver

service:
  extensions: [file_storage/filelogreceiver]
  pipelines:
    logs:
      receivers: [filelog]
      exporters: [...]
```

Assume there are no log files matching `/tmp/*.log` when the above collector pipeline starts executing. In this
scenario, the `/tmp/otelcol/file_storage/filelogreceiver` directory contains one file:

```
$ ls /tmp/otelcol/file_storage/filelogreceiver
receiver_filelog_
```

This is a binary file, so we can read its contents using the `strings` utility.

```
$ strings /tmp/otelcol/file_storage/filelogreceiver/receiver_filelog_
default
file_input.knownFiles0
default
file_input.knownFiles0
default

> file_input.knownFiles0

Author: @djaglowski Do you happen to know why the entries in the `filestorage` file are repeated 3 times? Best that I can tell, this has to do with how bbolt is storing the data on disk. I'd like to mention the reason here in these docs so users don't get confused by the repeated entries.

Reviewer 3: Interesting observation. I don't see any obvious reason from the fileconsumer side. Might be a bbolt peculiarity or a real problem with the storage extension or interface.

```

> When a new log file is created with a single entry in it, and the `filelog` receiver in the collector
> pipeline has ingested this entry, the contents of the `/tmp/otelcol/file_storage/filelogreceiver/receiver_filelog_` file
> change to reflect this new state.

Author: Fair point.  How about the following?
```suggestion
When a new log file is created with one or more entries in it, and the `filelog` receiver in the collector
pipeline has ingested these entries, the contents of the `/tmp/otelcol/file_storage/filelogreceiver/receiver_filelog_` file
change to reflect this new state.
```

> change to reflect this new state.

Reviewer 2: LGTM!
